File: receipt-service/app/models/receipt.py
"""Domain model for a fully validated, normalized receipt extraction.

Everything here has already passed through app.validation — nothing on this
model is a raw, untrusted LLM claim.
"""

from pydantic import BaseModel, Field


class LineItem(BaseModel):
    description: str
    quantity: float | None = None
    unit_price: float | None = None
    total_price: float | None = None


class Warning(BaseModel):
    field: str
    message: str
    severity: str = "warning"  # "warning" | "error"


class ExtractedReceipt(BaseModel):
    merchant: str | None = None
    expense_date: str | None = None  # ISO 8601 (YYYY-MM-DD)
    total: float | None = None
    subtotal: float | None = None
    tax: float | None = None
    currency: str | None = None  # ISO 4217, e.g. "EUR"
    # Always one of the caller-supplied category names, or None — never a
    # name the model invented (see app.validation.receipt_validator).
    category: str | None = None
    detected_language: str | None = None
    translation: str | None = None
    line_items: list[LineItem] = Field(default_factory=list)
    warnings: list[Warning] = Field(default_factory=list)
    raw_ocr_text: str = ""
    confidence: float = 0.0

File: types/receipt.ts
/** Mirrors receipt-service's ExtractedReceipt (receipt-service/app/models/receipt.py). */
export type ReceiptLineItem = {
  description: string;
  quantity: number | null;
  unit_price: number | null;
  total_price: number | null;
};

export type ReceiptWarning = {
  field: string;
  message: string;
  severity: "warning" | "error";
};

export type ExtractedReceipt = {
  merchant: string | null;
  expense_date: string | null;
  total: number | null;
  subtotal: number | null;
  tax: number | null;
  currency: string | null;
  detected_language: string | null;
  translation: string | null;
  line_items: ReceiptLineItem[];
  warnings: ReceiptWarning[];
  raw_ocr_text: string;
  confidence: number;
};
